gio/src/subclass/async_initable.rs

```diff
@@ -0,0 +1,522 @@
+            let res = AsyncInitableResult::new(
+                initable.as_ref(),
+                user_data,
+                async_initable_init_async::<T> as *mut _,
+                res.err(),
+            );
+            callback(
+                initable.upcast_ref::<glib::Object>().to_glib_none().0,
+                res.upcast_ref::<AsyncResult>().to_glib_none().0,
+                user_data,
+            );
+        }
+    });
+    imp.init_async(
+        initable.unsafe_cast_ref(),
+        from_glib(io_priority),
+        Option::<Cancellable>::from_glib_borrow(cancellable)
+            .as_ref()
+            .as_ref(),
+        callback,
+    );
+}
+
+unsafe extern "C" fn async_initable_init_finish<T: AsyncInitableImpl>(
+    initable: *mut ffi::GAsyncInitable,
+    res: *mut ffi::GAsyncResult,
+    error: *mut *mut glib::ffi::GError,
+) -> glib::ffi::gboolean {
+    let initable = from_glib_borrow::<_, AsyncInitable>(initable);
+    let res = from_glib_borrow::<_, AsyncResult>(res);
+
+    let res = res
+        .downcast_ref::<AsyncInitableResult>()
+        .expect("Result is not a Rust AsyncInitableResult");
+    if !res.is_valid(initable.as_ref()) {
+        panic!("AsyncInitable is not valid for source object");
+    }
+
+    match res.propagate_error() {
+        None => glib::ffi::GTRUE,
+        Some(e) => {
+            if !error.is_null() {
+                *error = e.into_raw();
+            }
+            glib::ffi::GFALSE
+        }
+    }
+}
+
+// A GAsyncResult that only holds the object and a GError, because GTask is too heavy
+glib::wrapper! {
+    struct AsyncInitableResult(ObjectSubclass<imp::AsyncInitableResult>)
+        @implements AsyncResult;
+}
+
+mod imp {
+    use crate::prelude::*;
+    use crate::subclass::prelude::*;
+    use crate::AsyncInitable;
+
+    impl super::AsyncInitableResult {
+        #[inline]
+        pub(super) fn new(
+            source_object: &AsyncInitable,
+            user_data: glib::Pointer,
+            source_tag: glib::Pointer,
+            error: Option<glib::Error>,
+        ) -> Self {
+            let res = glib::Object::new::<Self>(&[]).unwrap();
+            res.imp()
+                .inner
+                .set(Inner {
+                    source_object: source_object.clone(),
+                    user_data,
+                    source_tag,
+                })
+                .unwrap();
+            res.imp().error.replace(error);
+            res
+        }
+        #[inline]
+        pub(super) fn is_valid(&self, source_object: &AsyncInitable) -> bool {
+            self.imp().inner.get().unwrap().source_object == *source_object
+        }
+        #[inline]
+        pub(super) fn propagate_error(&self) -> Option<glib::Error> {
+            self.imp().error.take()
+        }
+    }
+
+    #[derive(Debug)]
+    struct Inner {
+        source_object: AsyncInitable,
+        user_data: glib::Pointer,
+        source_tag: glib::Pointer,
```

Comment:
This is not `Send` / `Sync`, which seems potentially problematic?

Reply:
I'm not sure, the purpose of this class is to not use threads at all. It only gets instantiated in the thread that finishes the async operation after it's over.